Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -113,10 +127,94 @@ inline void ComputeStrokeStyle(ID2D1DeviceContext* deviceContext) {
     inline void ClearStrokeStyle() {
         strokeStyle.Reset();
     }
+
+    inline bool HasShadow() {
+        return std::fpclassify(shadowColor.w) != FP_ZERO;
+    }
+
+    inline bool ShouldDraw() {
+        return std::fpclassify(alpha) != FP_ZERO && std::fpclassify(globalAlpha) != FP_ZERO;
+    }
+
+    inline HRESULT IntersectClippingGeometry(ID2D1Geometry* incomingGeometry, CGPathDrawingMode pathMode) {
+        D2D1_FILL_MODE d2dFillMode = (pathMode & kCGPathEOFill) == kCGPathEOFill ? D2D1_FILL_MODE_ALTERNATE : D2D1_FILL_MODE_WINDING;
+
+        if (!clippingGeometry) {
+            // If we don't have a clipping geometry, we are free to take this one wholesale (after EO/Winding conversion.)
+            return _CGConvertD2DGeometryToFillMode(incomingGeometry, d2dFillMode, &clippingGeometry);
+        }
+
+        ComPtr<ID2D1Factory> factory;
+        clippingGeometry->GetFactory(&factory);
+
+        // If we have a clipping geometry, we must intersect it with the new path.
+        // To do so, we need to stream the combined geometry into a totally new geometry.
+        ComPtr<ID2D1PathGeometry> newClippingPathGeometry;
+        RETURN_IF_FAILED(factory->CreatePathGeometry(&newClippingPathGeometry));
+
+        ComPtr<ID2D1GeometrySink> geometrySink;
+        RETURN_IF_FAILED(newClippingPathGeometry->Open(&geometrySink));
+
+        geometrySink->SetFillMode(d2dFillMode);
+
+        RETURN_IF_FAILED(clippingGeometry->CombineWithGeometry(incomingGeometry, D2D1_COMBINE_MODE_INTERSECT, nullptr, geometrySink.Get()));
+
+        RETURN_IF_FAILED(geometrySink->Close());
+
+        clippingGeometry.Attach(newClippingPathGeometry.Detach());
+        return S_OK;
+    }
+};
+
+class __CGContextLayer {
+    std::stack<__CGContextDrawingState> _stateStack;
+    ComPtr<ID2D1Image> _target;
+
+public:
+    __CGContextLayer(ID2D1Image* target, CGRect* region) : _target(target) {
+        // Each newly-pushed default-constructed layer has an empty base state.
+        _stateStack.emplace();
+    }
+
+    __CGContextLayer(ID2D1Image* target, CGRect* region, __CGContextLayer& parentLayer) : _target(target) {
+        _stateStack.emplace(parentLayer.CurrentGState());
+    }
+
+    __CGContextDrawingState& CurrentGState() {
+        return _stateStack.top();
+    }
+
+    // PushGState saves the provided Direct2D drawing state into the current rendering
+    // configuration and creates a new empty one.
+    inline HRESULT PushGState(ID2D1DrawingStateBlock* drawingState) {
+        RETURN_HR_IF(E_INVALIDARG, !drawingState);
+
+        auto& oldState = _stateStack.top();
+        _stateStack.emplace(oldState);
```

Comment:
Yes?

---
In reply to: [90588464](https://github.com/Microsoft/WinObjC/pull/1496#discussion_r90588464) [](ancestors = 90588464)